File: 2020/1/ruby/1.rb
test_data = [
  1721,
  979,
  366,
  299,
  675,
  1456,
]

def search_part1(data, target)
  data.permutation(2).to_a.each { |a, b|
    return [a, b].inject(:*) if a + b == target
  }
end

def search_part2(data, num_factors, target)
  data.permutation(num_factors).to_a.each { |a|
    return a.inject(:*) if a.inject(:+) == target
  }
end

product = search_part1(test_data, 2020)
raise "Invalid test" unless product == 514579

input_file = File.join(File.dirname(__FILE__), "../input")
part1_data = File.readlines(input_file).map(&:to_i)

product = search_part1(part1_data, 2020)
puts "Part 1 result: #{product}"

product = search_part2(part1_data, 3, 2020)
puts "Part 2 result: #{product}"

File: 2020/1/python/1.py
import os
from pathlib import Path
from itertools import combinations
from functools import reduce

test_data = [
    1721,
    979,
    366,
    299,
    675,
    1456,
]


def search(data, num_factors=2, target=2020):
    for values in combinations(sorted(data), num_factors):
        if sum(values) == target:
            return reduce(lambda x, y: x * y, values)


if __name__ == "__main__":
    filename = Path(os.path.dirname(__file__)) / "../input"
    with filename.open() as f:
        real_data = [int(i) for i in f]

    assert search(test_data, target=2020) == 514579
    print("Part 1: ", search(real_data, target=2020))
    print("Part 2: ", search(real_data, target=2020, num_factors=3))
